File: 11_动态规划/dp优化/辅助数据结构dp/单调队列优化dp/MonoQueue.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */

/**
 * 单调队列维护滑动窗口最小值.
 */
class MonoQueue<T> {
  minQueue: SimpleQueue<T> = new SimpleQueue()
  private _len = 0
  private readonly _minQueueCount: SimpleQueue<number> = new SimpleQueue()
  private readonly _compareFn: (a: T, b: T) => number

  constructor(compareFn: (a: T, b: T) => number) {
    this._compareFn = compareFn
  }

  append(value: T): this {
    let count = 1
    while (this.minQueue.length && this._compareFn(this.minQueue.back()!, value) > 0) {
      this.minQueue.pop()
      count += this._minQueueCount.pop()!
    }
    this.minQueue.append(value)
    this._minQueueCount.append(count)
    this._len++
    return this
  }

  popleft(): void {
    if (!this._len) return
    this._minQueueCount.set(0, this._minQueueCount.front()! - 1)
    if (!this._minQueueCount.front()!) {
      this.minQueue.popleft()
      this._minQueueCount.popleft()
    }
    this._len--
  }

  head(): T | undefined {
    return this.minQueue.front()
  }

  toString(): string {
    const res: string[] = []
    for (let i = 0; i < this.minQueue.length; ++i) {
      res.push(JSON.stringify({ value: this.minQueue.at(i), count: this._minQueueCount.at(i) }))
    }
    return `MonoQueue[${res.join(', ')}]`
  }

  get min(): T | undefined {
    return this.minQueue.front()
  }

  get length(): number {
    return this._len
  }
}

class SimpleQueue<T> {
  private readonly _queue: T[] = []
  private _left = 0

  append(value: T): this {
    this._queue.push(value)
    return this
  }

  pop(): T | undefined {
    return this._queue.pop()
  }

  popleft(): T | undefined {
    if (!this.length) return undefined
    return this._queue[this._left++]
  }

  at(index: number): T | undefined {
    return this._queue[index + this._left]
  }

  set(index: number, value: T): void {
    this._queue[index + this._left] = value
  }

  toString(): string {
    const res: string[] = []
    for (let i = 0; i < this.length; ++i) {
      res.push(JSON.stringify(this.at(i)))
    }
    return `[${res.join(', ')}]`
  }

  front(): T | undefined {
    return this._queue[this._left]
  }

  back(): T | undefined {
    return this._queue[this._queue.length - 1]
  }

  get length(): number {
    return this._queue.length
  }
}

export { MonoQueue }

if (require.main === module) {
  const queue = new SimpleQueue<number>()
  queue.append(1)
  console.log(queue.toString())
  const monoQueue = new MonoQueue<number>((a, b) => a - b)
  monoQueue.append(2)
  monoQueue.append(1)
  console.log(monoQueue.toString())
  console.log(monoQueue.minQueue.toString())

  // https://leetcode.cn/problems/minimum-number-of-coins-for-fruits/description/
}

File: 11_动态规划/dp优化/辅助数据结构dp/单调队列优化dp/MonoQueue.go
package main

import (
	"fmt"
	"strings"
)

// https://leetcode.cn/problems/minimum-number-of-coins-for-fruits/description/
func main() {
	queue := NewMonoQueue(func(a, b V) bool { return a.v < b.v })
	queue.Append(V{1, 1})
	fmt.Println(queue)
}

type V = struct{ v, i int }

// 单调队列维护滑动窗口最小值.
// 单调队列队头元素为当前窗口最小值，队尾元素为当前窗口最大值.
type MonoQueue struct {
	MinQueue       []V
	_minQueueCount []int32
	_less          func(a, b V) bool
	_len           int
}

func NewMonoQueue(less func(a, b V) bool) *MonoQueue {
	return &MonoQueue{
		_less: less,
	}
}

func (q *MonoQueue) Append(value V) *MonoQueue {
	count := int32(1)
	for len(q.MinQueue) > 0 && q._less(value, q.MinQueue[len(q.MinQueue)-1]) {
		q.MinQueue = q.MinQueue[:len(q.MinQueue)-1]
		count += q._minQueueCount[len(q._minQueueCount)-1]
		q._minQueueCount = q._minQueueCount[:len(q._minQueueCount)-1]
	}
	q.MinQueue = append(q.MinQueue, value)
	q._minQueueCount = append(q._minQueueCount, count)
	q._len++
	return q
}

func (q *MonoQueue) Popleft() {
	q._minQueueCount[0]--
	if q._minQueueCount[0] == 0 {
		q.MinQueue = q.MinQueue[1:]
		q._minQueueCount = q._minQueueCount[1:]
	}
	q._len--
}

func (q *MonoQueue) Head() V {
	return q.MinQueue[0]
}

func (q *MonoQueue) Min() V {
	return q.MinQueue[0]
}

func (q *MonoQueue) Len() int {
	return q._len
}

func (q *MonoQueue) String() string {
	sb := []string{}
	for i := 0; i < len(q.MinQueue); i++ {
		sb = append(sb, fmt.Sprintf("%v", pair{q.MinQueue[i], q._minQueueCount[i]}))
	}
	return fmt.Sprintf("MonoQueue{%v}", strings.Join(sb, ", "))
}

type pair struct {
	value V
	count int32
}

func (p pair) String() string {
	return fmt.Sprintf("(value: %v, count: %v)", p.value, p.count)
}
